synapse/handlers/federation.py

```diff
@@ -2711,9 +2711,11 @@ async def _update_auth_events_and_context_for_auth(
                             event.event_id,
                             e.event_id,
                         )
-                        context = await self.state_handler.compute_event_context(e)
+                        missing_auth_event_context = (
```

Comment:
Is there any linting we can add to avoid re-assigning or mask existing variables?

Reply:
@erikjohnston Still curious if there is some tooling we can use to avoid this problem?

Since Python is not block scoped, I think we do this all over on purpose but would be nice to be more conscious about it.

Reply:
From some cursory googling around Python linting (black, flake8, mypy, type hinting, etc), I'm not finding anything around constants, declarations vs assignments, and re-assignments 😥

Reply:
It looks like there is a handful of related violations in the `wemake-python-styleguide` best practices section. But none of these will specifically stop function parameter re-assignment for this exact bug scenario (tried it in a code playground).

 - [`BlockAndLocalOverlapViolation`](https://wemake-python-stylegui.de/en/latest/pages/usage/violations/best_practices.html#wemake_python_styleguide.violations.best_practices.BlockAndLocalOverlapViolation): Forbid overlapping local and block variables.
 - [`OuterScopeShadowingViolation`](https://wemake-python-stylegui.de/en/latest/pages/usage/violations/best_practices.html#wemake_python_styleguide.violations.best_practices.OuterScopeShadowingViolation): Forbid shadowing variables from outer scopes.
 - [`ControlVarUsedAfterBlockViolation`](https://wemake-python-stylegui.de/en/latest/pages/usage/violations/best_practices.html#wemake_python_styleguide.violations.best_practices.ControlVarUsedAfterBlockViolation): Forbid control variables after the block body.



### Dev notes


Interesting comparison of linting type tools: https://github.com/wemake-services/wemake-python-styleguide#what-we-are-about

---

Related but not related links (just linking for my own reference):

 - https://github.com/satwikkansal/wtfPython#-the-disappearing-variable-from-outer-scope
 - https://github.com/satwikkansal/wtfPython#-the-out-of-scope-variable
 - https://github.com/satwikkansal/wtfPython#-loop-variables-leaking-out
 - https://github.com/satwikkansal/wtfPython#-name-resolution-ignoring-class-scope

Reply:
I created https://github.com/wemake-services/wemake-python-styleguide/issues/2128 to propose a new linting rule for this scenario 🐣

Reply:
We could use [Semgrep](https://github.com/returntocorp/semgrep) to find all the scenarios in the codebase where we do this.